feat(windrose): Wind Rose GH component and sample script.

saeranv opened the pull request:
@chriswmackey 
cc @mostaphaRoudsari 

This PR contains the GH component for the Wind Rose. It includes a GH script in the sample folder with an example of the wind rose. 

![image](https://user-images.githubusercontent.com/1685956/85364052-4fab5280-b4f0-11ea-8f64-c8a3ea0d10fa.png)

![image](https://user-images.githubusercontent.com/1685956/85364079-5b971480-b4f0-11ea-93cb-1c4b901af24f.png)

I think I've gotten it working so it has all the functionality you would want. There is one bug that I've encountered, which is that when the show_calmrose option is selected, the calmrose mesh is slightly smaller then the grid lines (which are correct). I've looked into this and I believe it's an issue with the floating point precision interfering with scaling. I'm still working on a fix, but think the rest of this component is ready for review.  

Also, apologies for not assigning reviewers/assignees in this PR, I don't think I have the permissions for it.

chriswmackey commented on ladybug_grasshopper/src/LB Wind Rose Plot.py:
```diff
@@ -0,0 +1,176 @@
+            difference between the North and the positive Y-axis in degrees.
+            90 is West and 270 is East. This can also be Vector for the
+            direction to North. (Default: 0)
+        _wind_speed_data: A HourlyContinuousCollection or HourlyDiscontinuousCollection of wind
```
We have been trying to keep the names of the component inputs shorter in [+] if only to avoid really big components. And, instead of packing a lot of words into the variable name, we just write good descriptions. Can we change this to just `_wind_speed`?

chriswmackey commented on ladybug_grasshopper/src/LB Wind Rose Plot.py:
```diff
@@ -0,0 +1,176 @@
+        _wind_speed_data: A HourlyContinuousCollection or HourlyDiscontinuousCollection of wind
+            values, corresponding to the wind directions, which is "binned" by the
+            calculated direction intervals.
+        _wind_direction_data: A HourlyContinuousCollection or
```
... and this to `_wind_direction`

chriswmackey commented on ladybug_grasshopper/src/LB Wind Rose Plot.py:
```diff
@@ -0,0 +1,176 @@
+        _wind_direction_data: A HourlyContinuousCollection or
+            HourlyDiscontinuousCollection of wind directions which will be used to "bin"
+            the intervals for the windrose.
+        _number_of_directions_: Number that determines the number of directions to the
```
... and this to `_direction_count_` or even `_dir_count_`

chriswmackey commented on ladybug_grasshopper/src/LB Wind Rose Plot.py:
```diff
@@ -0,0 +1,176 @@
+            in time while the color illustrates the average _wind_speed_data values
+            associated with each interval. The number of frequency lines with similar
+            colors therefore indicate a higher frequency of that value.
+        _frequency_distance_: The distance for the frequency interval in model units
```
... and this to `_freq_dist_`

chriswmackey commented on ladybug_grasshopper/src/LB Wind Rose Plot.py:
```diff
@@ -0,0 +1,176 @@
+            corresponding to the total zero values divided by the number of directions.
+            This allows the wind rose to represent zero values from _wind_speed_data
+            even though such values don't have any direction associated with them.
+        _show_frequency_: A boolean to show the frequency of _wind_speed_data values
```
... and this to `_show_freq_`

chriswmackey commented on ladybug_grasshopper/src/LB Wind Rose Plot.py:
```diff
@@ -0,0 +1,176 @@
+
+    Returns:
+        report: ...
+        wind_histogram_data: A Grasshopper data tree with the number of branches equal to the
```
... and this to hist_data

chriswmackey commented on ladybug_grasshopper/src/LB Wind Rose Plot.py:
```diff
@@ -0,0 +1,176 @@
+        mesh: A colored mesh representing the wind rose derived from the input data.
+        compass: A set of circles, lines and text objects that mark the cardinal
+            directions in relation to the wind rose.
+        orientation_lines: Line geometries representing the edges (or "spokes") of the wind rose
```
... and this to `orient_lines`

chriswmackey commented on ladybug_grasshopper/src/LB Wind Rose Plot.py:
```diff
@@ -0,0 +1,176 @@
+            directions in relation to the wind rose.
+        orientation_lines: Line geometries representing the edges (or "spokes") of the wind rose
+            directions.
+        frequency_lines: Polygon geometries representing the frequency intervals of the wind rose.
```
... and this to `freq_lines`

chriswmackey commented on ladybug_grasshopper/src/LB Wind Rose Plot.py:
```diff
@@ -0,0 +1,176 @@
+            (Default: 10). If _show_calmrose is set to True, then the initial frequency
+            interval corresponds to the number of calm hours in the data collection,
+            which may not align with this _frequency_spacing.
+       _frequency_hours_: The number of hours represented in each frequency interval
```
And this to `_freq_hours_`

chriswmackey commented on ladybug_grasshopper/src/LB Wind Rose Plot.py:
```diff
@@ -0,0 +1,176 @@
+            which may not align with this _frequency_spacing.
+       _frequency_hours_: The number of hours represented in each frequency interval
+            (Default: 200).
+        _max_frequency_:A number representing the maximum hours that the outermost ring
```
... and this to `_max_freq_`

chriswmackey commented on ladybug_grasshopper/src/LB Wind Rose Plot.py:
```diff
@@ -0,0 +1,176 @@
+            frequency intervals in the wind rose. If nothing is provided, a default
+            LegendParameter object is computed using values from the wind data with
+            11 segments (Default: None).
+        period_: A Ladybug analysis period to be applied to all of the input _data.
```
I see that there's no input for additional data or a conditional statement. If you could add these in, I think it would be really valuable and you can see [here that the application of a conditional statement should only take 3 lines of code](https://github.com/ladybug-tools/ladybug-grasshopper/blob/master/ladybug_grasshopper/src/LB%20Hourly%20Plot.py#L88-L90)

I am also happy to take care of this part if you could just give me some sample code for how you would generate a colored wind rose mesh like so:
![image](https://user-images.githubusercontent.com/5567574/85416282-45af4100-b53c-11ea-9499-77da3b244c9d.png)

saeranv replied:
Thanks @chriswmackey for the offer. I agree it looks fairly simple, so I think I can implement it with my next PR. I'll let you know if I run into any trouble.

chriswmackey replied:
Thanks, @saeranv . That will be a big help.

saeranv commented:
@chriswmackey 

I've changed the docstrings, and also added the ability for the component to handle multiple data collections. 

However, I ran into trouble in two areas that I need some advice on: 

1. The conditional statement_ seems to only work on one data collection, but in this case we have two (wind speed and direction). Is there a way I can filter the wind speed values with the conditional statement, - while also removing the same values from the wind direction collection so that they remain equivalent (i.e returning indices that were kept to apply to another collection)? Right now (from the example script I'm referencing in the hourly_plot component) I can only filter one data collection, which results in the wind speed values being a different length from the wind direction values.

2. Since we have two data collections, I was thinking of adding two statement_ inputs (i.e `speed_statement_` and `dir_statement_`), so that the user can modify each. If there's a way to separately filter both of them with one statement argument, let me know.

saeranv commented on ladybug_grasshopper/src/LB Wind Rose Plot.py:
```diff
@@ -0,0 +1,258 @@
+    if len(_wind_speed) > len(legend_par_) > 0:
+        num = len(_wind_speed) - len(legend_par_)
+        legend_par_ += [legend_par_[-1] for i in range(num)]
+
```
@chriswmackey 

I think I might delete the option to include multiple legends and just change the input to 'item' type.  I've noticed that when I duplicate legends I seem to lose the min/max bounds, and the option for multiple legends seems a little unnecessary given that legends are automatically generated for specific bounds if you leave the input blank, so the only use case for wanting to control them will be if you want to control all of them at once. Thoughts?

chriswmackey replied:
@saeranv ,
Now that I understand this better, I think that the `_wind_direction` data should be item access and the `_wind_data` should be list access. So you can plot multiple data streams with the same wind direction data.

chriswmackey commented:
Thanks @saeranv ,

All that you have to do to make sure that the wind direction and wind speed are aligned with the other data collections is include it as the last item in the list you are inputting into the `filter_collections_by_statement` method. So, you do exactly what you see here:
https://github.com/ladybug-tools/ladybug-grasshopper/blob/master/ladybug_grasshopper/src/LB%20Hourly%20Plot.py#L91-L93
... except `_data` should be a list with the extra input `_data` to the component first (eg. dry bulb temperature, relative humidity). Then, you append the wind speed and wind direction to this list.

Also, I think everything into this component should be of the Item grasshopper type except `_data`, which should be for any number of additional data collections that are aligned with the wind speed and wind direction.

I also think that we should delete the `show_freq` input, If people don't want to see the frequency output, then they can do this with native Grasshopper preview capabilities.

chriswmackey commented:
Also, for clarification, do you have a method in the core libraries to get the wind rose mesh colored with data other than wind speed (eg. with temperature data like I had posted)? I was looking through your core library code and I couldn't find anything that specifically referenced it.

chriswmackey commented:
Ah wait, I think I didn't realize that your input for [`analysis_data_collection`](https://github.com/ladybug-tools/ladybug/blob/master/ladybug/windrose.py#L32-L34) can be of any data type (not just wind speed). Is that correct?

If so, I definitely support treating any component list access with the assumption that all data collections are aligned (eg. have passed through the same conditional statement if there is a statement input). So, if there is access for a list of legend parameters, it is there to assign different legend parameters to different data types (eg. temperature, humidity) across which the data is applied. Not to assign it to multiple inputs of wind speed or wind direction.

saeranv commented:
@chriswmackey 

Yes that's right, it can be of any data type. I'll change it the name to wind_data_ to make that more clear.

Re: show_freq: This parameter is not just to show the preview lines, but actually calculates the wind speed per frequency interval as well. So if you don't select it, it just shows the average data per direction:

![image](https://user-images.githubusercontent.com/1685956/86043774-28d0ad00-ba17-11ea-81b5-d99ef53e7763.png)

saeranv commented:
@chriswmackey 

Thanks, got it working!

chriswmackey commented on ladybug_grasshopper/src/LB Wind Rose Plot.py:
```diff
@@ -0,0 +1,269 @@
+        for i in range(len(_wind_data)):
+            _data = HourlyContinuousCollection.filter_collections_by_statement(
+                [_wind_data[i], _wind_direction[i]], statement_)
+            _wind_data[i], _wind_direction[i] = _data
```
This should be:
```
_data = HourlyContinuousCollection.filter_collections_by_statement(
                _wind_data + [_wind_direction], statement_)
_wind_data = _data[:-1]
_wind_direction = _data[-1]
```

saeranv replied:
Yes, if there's only one _wind_direction item this makes more sense.

chriswmackey commented on ladybug_grasshopper/src/LB Wind Rose Plot.py:
```diff
@@ -0,0 +1,269 @@
+            assert -360.0 <= north_ <= 360.0, 'The north orientation must be greater ' \
+                'then -360 and less then 360 to plot the wind rose. ' \
+                'Got: {}'.format(north_)
+            north_ = float(north_)
```
It doesn't make sense for you to do type-casting after you do a check that requires the input to be a float. Put the assert statement after the type casting. Also, don't you already handle this case in the core libraries with the `% 360` operator? So you don't need the assert?

saeranv replied:
@chriswmackey 
Yes re: type-check/cast. 

The reason I'm checking that it's between -360 and 360 here is because the sun plot component is also checking those constraints for it's north_ input. So I thought it would be better to keep our north_ inputs consistent for the GH components.

chriswmackey commented on ladybug_grasshopper/src/LB Wind Rose Plot.py:
```diff
@@ -0,0 +1,269 @@
+            max_freqs.append(w.frequency_intervals_compass)
+        _max_freq_lines_ = max(max_freqs)
+
+    for i, (dir_data, speed_data) in enumerate(zip(_wind_direction, _wind_data)):
```
This should get a lot simpler when you are only iterating through the `_wind_data`

chriswmackey commented on ladybug_grasshopper/src/LB Wind Rose Plot.py:
```diff
@@ -0,0 +1,269 @@
+        if _show_freq_:
+            freq_lines = [from_polygon2d(poly) for poly in windrose.frequency_lines]
+
+        # move the center point so sun paths are not on top of one another
```
There are no "sun paths" in this component 😉

chriswmackey commented on ladybug_grasshopper/src/LB Wind Rose Plot.py:
```diff
@@ -0,0 +1,269 @@
+        _wind_data: A HourlyContinuousCollection or HourlyDiscontinuousCollection of
+            wind values, corresponding to the wind directions, which is "binned" by the
+            calculated direction intervals. This input usually consists of wind speed
+            values, but is not limited to this data type.
```
You should change this description to note that it accepts "A list of data collections" just as easily as it accepts a single collection. I also think that we should probably just call it `_data` instead of `_wind_data`. If the goal is to show users that users can plug in any data collection they want, it's better to use the more abstract term.

saeranv commented:
@chriswmackey 

This should handle all the revisions noted. 

Also, I had to make another change to the ladybug/windrose.py, since I realized that the windrose was filtering out zero-values (and showing it only in the calmrose), which isn't a technique that makes sense if we allow non-speed data types to be plotted.

chriswmackey commented:
Looks good @saeranv . I will merge tomorrow at a time where this won't create conflicts with the other libraries.
